### PR TITLE
fix(receive): Rewrite gateway message queue logic

### PR DESCRIPTION
**Pull request by ettersi**
Fixes https://github.com/org-arl/Fjage.jl/issues/30. Alternative to https://github.com/org-arl/Fjage.jl/pull/32.

### CODE REVIEW

**Comment by ettersi** [comment]
I added [some benchmarks](https://github.com/org-arl/Fjage.jl/blob/d71576c540df8597a755a32c98c93007414213ff/bench/bench.jl) as suggested during our discussion on Friday. I'll report on my findings below, but keep in mind that benchmarking is tricky and so everything should be taken with a good grain of salt. (There's an awesome [German saying](https://translate.google.com/?sl=de&tl=en&text=Wer%20misst%2C%20misst%20Mist.%20&op=translate) for this.)

The first benchmark measures the time for pushing an popping a message into and out of an otherwise empty message queue. With the current implementation, this takes about 500ns. 

```
------------------------
Gateway, send -> receive
------------------------

BenchmarkTools.Trial: 10000 samples with 190 evaluations.
 Range (min … max):  519.668 ns …  12.756 μs  ┊ GC (min … max): 0.00% … 93.96%
 Time  (median):     538.479 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   574.685 ns ± 419.554 ns  ┊ GC (mean ± σ):  3.01% ±  3.96%

 Memory estimate: 448 bytes, allocs estimate: 9.
```

500ns should be more than acceptable, but out of curiosity I decided to also benchmark Julia's `Channel`s implementation on a similar operation. Turns out we are only about 5-6 times slower than that. 

```
-----------------------------------------
Channel, send -> receive (for comparison)
-----------------------------------------

BenchmarkTools.Trial: 10000 samples with 943 evaluations.
 Range (min … max):   95.150 ns …  1.900 μs  ┊ GC (min … max): 0.00% … 93.98%
 Time  (median):     100.225 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   103.435 ns ± 39.347 ns  ┊ GC (mean ± σ):  0.78% ±  2.07%

 Memory estimate: 32 bytes, allocs estimate: 1.
```

------

I expect that in most real-world scenarios, the message queue is full rather than empty, so to make the above benchmark more realistic, I tweaked it such that it pushes and pops the last message out of a full rather than an empty message queue. This increases the runtime quite significantly to about 25µs. 

```
------------------------------------
Gateway, send -> receive, full queue
------------------------------------

BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  22.011 μs …  1.676 ms  ┊ GC (min … max): 0.00% … 97.35%
 Time  (median):     22.990 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   25.142 μs ± 27.483 μs  ┊ GC (mean ± σ):  1.84% ±  1.69%

 Memory estimate: 16.41 KiB, allocs estimate: 520.
```

Some further investigations (not reflected in this code) indicate that this is because iterating over an array of references is quite slow in Julia (about 100ns per entry). 

----

Next, I switched the order of the operations so the pull happens before the push. Once again, this significantly increases the runtime to about 30µs. 

```
------------------------
Gateway, receive -> send
------------------------

BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  29.087 μs … 138.687 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     33.453 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   36.441 μs ±   8.331 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

 Memory estimate: 480 bytes, allocs estimate: 10.
```

It seems that almost all of this runtime is due to switching between tasks. To back this up, I once again also benchmarked Julia's `Channel`s on an equivalent operation, and I also benchmarked a ping-pong using `Event`s (which I believe is the lowest-overhead solution to switch back and forth between two tasks). These alternatives take exactly as long as our gateway code. 

```
-----------------------------------------
Channel, receive -> send (for comparison)
-----------------------------------------

BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  27.725 μs … 214.435 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     31.262 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   33.778 μs ±   8.092 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

 Memory estimate: 32 bytes, allocs estimate: 1.


--------------------------------
Event ping pong (for comparison)
--------------------------------

BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  28.377 μs …   5.945 ms  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     34.080 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   39.510 μs ± 106.370 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

 Memory estimate: 0 bytes, allocs estimate: 0.
```

---

Finally, I repeated the above benchmark but added two more receivers which just sit idly in the receive queue. The conclusion here is that handling each additional receiver adds about 30µs to the overall runtime. 

```
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):   86.259 μs … 336.188 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     105.255 μs               ┊ GC (median):    0.00%
 Time  (mean ± σ):   112.813 μs ±  23.615 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

 Memory estimate: 928 bytes, allocs estimate: 24.
```

We can put this number into perspective by doing some quick back-of-the-envelope calculations: if we give ourselves a time budget of 1ms to deal with message delivery, then 30µs per receiver means that this implementation can support about 30 active agents. I guess that's good enough for now, but also not very far from potentially troublesome territory.

**Comment by ettersi** [comment]
I've also added a [simple stresstest](https://github.com/org-arl/Fjage.jl/blob/gw-receive-v2/test/stresstest/stresstest.jl) which does the following:

- A single sender spams the message queue as fast as it can. 
- Five receivers repeatedly and randomly pick one of the following operations. 
   1) Accept the `n`th message that is presented to you, where `n = rand(1:2*Fjage.MAX_QUEUE_LEN)`. 
   2) Decline all messages and just wait for the time out. 

The key goal here is to ensure that the system neither crashes nor deadlocks. I'm keeping track of the latter by monitoring

1) the time it takes to receive messages (top axis), and
2) the timeout accuracy (bottom axis). 
 
stresstest
 
Both plots look exactly as expected except for a few outliers. My best guess is that these outliers are due to garbage collection or scheduling jitter. In either case, I don't think they are a cause for concern. 

--- 

This branch should now be ready for merging as far as I can tell. Please let me know in case I've missed anything.

**Comment by ettersi** [comment]
The test failure is due to a very old version of Julia ([v1.3](https://github.com/org-arl/Fjage.jl/blob/dc74a5edff847e00433e2ac05fde96c5c4b8c4e4/.github/workflows/main.yml#L19)). If this needs to be resolved, maybe we can resolve it by upgrading Julia?

File: test/stresstest/Project.toml
Comment (mchitre) [inline test/stresstest/Project.toml:3]: Likely to have problems with CI with GLMakie? Do we need it for tests?

File: test/stresstest/stresstest.jl
Comment (mchitre) [inline test/stresstest/stresstest.jl:104]: While plotting is nice for human checking, perhaps not needed in CI. Maybe have it on some kind of manual mode when human debugging?

**Comment by mchitre** [comment]
> The test failure is due to a very old version of Julia ([v1.3](https://github.com/org-arl/Fjage.jl/blob/dc74a5edff847e00433e2ac05fde96c5c4b8c4e4/.github/workflows/main.yml#L19)). If this needs to be resolved, maybe we can resolve it by upgrading Julia?

We can move to Julia 1.6 LTS for tests.

**Review by mchitre — APPROVED** [review]
In principal approved, provided the concerns raised above are addressed. No need for re-approval from me, unless @ettersi needs specific inputs.

File: test/stresstest/Project.toml
Comment (ettersi) [reply test/stresstest/Project.toml:3]: This is in a separate `test/ >>> stresstest <<< /Project.toml` file, not `test/Project.toml`. So this won't be loaded as part of CI.

File: test/stresstest/stresstest.jl
Comment (ettersi) [reply test/stresstest/stresstest.jl:104]: `stresstest.jl` is not loaded as part of `runtests.jl`, so this entire file is restricted to "manual mode". The flip side of this is that the stress test is not part of the usual CI. It is not yet clear to me how to write an automatic test suite which is both reasonably strict but also not susceptible to false positives, so I was thinking maybe we could leave this as is until I have more experience / more ideas? I'm also open to suggestions...

**Review by notthetup — APPROVED** [review]

**Pull request merged**
